vm/object.c

```diff
@@ -64,27 +64,49 @@ int vm_objectGet(vm_object_t **o, oid_t oid)
 	*o = lib_treeof(vm_object_t, linkage, lib_rbFind(&object_common.tree, &t.linkage));
 
 	if (*o == NULL) {
-		sz = (size_t)proc_size(oid);
-		n = round_page(sz) / SIZE_PAGE;
+		/* Take off the lock to avoid a deadlock in vm_kmalloc */
+		proc_lockClear(&object_common.lock);
 
-		if ((*o = (vm_object_t *)vm_kmalloc(sizeof(vm_object_t) + n * sizeof(page_t *))) == NULL) {
-			proc_lockClear(&object_common.lock);
+		sz = proc_size(oid);
+		if (sz < 0) {
+			return (int)sz;
```

Comment:
Hmm, should we check again for object after `proc_size` fail (it calls `vm_kmalloc`)? This should probably apply to all error paths.

Reply:
I don't know, this really shouldn't happen

Reply:
It's very unlikely but it could happen. You could add `else` to `if` at lines 71 and 74 and nest them so that in case of failure we'll always proceed to second object check. Error code could be saved in a new variable and returned at line 87.